registry/storage/layerreader.go

```diff
@@ -73,7 +74,21 @@ func (lr *layerReader) Handler(r *http.Request) (h http.Handler, err error) {
 	}
 
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// If the registry is serving this content itself, check
+		// the If-None-Match header and return 304 on match.  Redirected
+		// storage implementations do the same.
+
+		for _, etag := range r.Header["If-None-Match"] {
+			dgst := lr.Digest()
+			if etag == dgst.String() {
+				w.WriteHeader(http.StatusNotModified)
+				return
+			}
+		}
+
 		w.Header().Set("Docker-Content-Digest", lr.digest.String())
+		w.Header().Set("ETag", lr.digest.String())
+		w.Header().Set("Cache-Control", fmt.Sprintf("max-age=%d", 86400))
```

Comment:
Yes. Does that make sense here? The issue is that a layer requires access control to the repository. If we allow CDNs to cache it without taking that access control into account, could we expose private data?

This may not matter since everything would be over https. As long as the SSL is terminated as the registry, this is probably okay.